src/masternodes/masternodes.cpp

```diff
@@ -576,10 +595,34 @@ Res CRewardsHistoryStorage::AddBalance(CScript const & owner, DCT_ID poolID, uin
 bool CRewardsHistoryStorage::Flush()
 {
     if (acindex) {
+        auto wallets = GetWallets();
         for (const auto& diff : diffs) {
+            bool isMine = false;
             const auto& pair = diff.first;
-            SetRewardHistory({pair.first, height, pair.second}, diff.second);
+            for (auto wallet : wallets) {
+                if ((isMine = IsMineCached(*wallet, pair.first) == ISMINE_SPENDABLE)) {
+                    SetMineRewardHistory({pair.first, height, pair.second}, diff.second);
+                    break;
+                }
+            }
+            if (!isMine && acindex == accountHistoryType::Full) {
+                SetAllRewardHistory({pair.first, height, pair.second}, diff.second);
+            }
         }
     }
     return CCustomCSView::Flush();
 }
+
+isminetype IsMineCached(CWallet const & wallet, CScript const & script)
+{
+    static std::unordered_map<std::string, std::map<CScript, isminetype>> mineCached;
```

Comment:
`mineCached` in rpc calls (`mn_rpc.cpp`), `ApplyCustomTx` and `ConnectBlock` is guarded by `cs_main` we can marked as `GUARDED_BY` it's limited no clang only.